PGM Crash Analysis

stwrt opened the pull request:
#### f209607bc69672c97760e70b3e128cc79d4a704d
<pre>
PGM Crash Analysis
<a href="https://bugs.webkit.org/show_bug.cgi?id=255345">https://bugs.webkit.org/show_bug.cgi?id=255345</a>

Reviewed by Mark Lam, Elliott Williams and David Kilzer.

When a process crashes we will now try to determine whether this crash was caused by
a PGM allocation in WebKit. ReportCrash will call into JSC, which would forward the information
onto libpas. libpas will respond with a report generated from the memory of the now dead process.

libpas will determine whether this was a PGM crash and if so what kind of crash (UAF or OOB).
This information will be added to the local crash log generated.

* Source/JavaScriptCore/API/PASReportCrashPrivate.cpp: Added.
(PASReportCrashExtractResults):
* Source/JavaScriptCore/API/PASReportCrashPrivate.h: Added.
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/SourcesCocoa.txt:
* Source/bmalloc/CMakeLists.txt:
* Source/bmalloc/bmalloc.xcodeproj/project.pbxproj:
* Source/bmalloc/libpas/libpas.xcodeproj/project.pbxproj:
* Source/bmalloc/libpas/src/libpas/pas_enumerate_large_heaps.c:
(pas_hash_map_entry_callback):
* Source/bmalloc/libpas/src/libpas/pas_report_crash.c: Added.
(memory_reader_adapter):
(setup_memory_reader):
(pas_report_crash_extract_pgm_failure):
* Source/bmalloc/libpas/src/libpas/pas_report_crash.h: Added.
* Source/bmalloc/libpas/src/libpas/pas_report_crash_pgm_report.h: Added.

Canonical link: <a href="https://commits.webkit.org/263055@main">https://commits.webkit.org/263055@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8bb5b510fa07f05259724828b9299003a3990d5c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/3478 "29 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/3529 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/3663 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/4900 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/3766 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/3612 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/3566 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/3020 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/3520 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/3119 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/4720 "Built successfully") | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/3098 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/3005 "Passed tests") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/2847 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/3075 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/3157 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/4466 "267 api tests failed or timed out") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/3260 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/3540 "Built successfully") | [⏳ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/macOS-AppleSilicon-Ventura-Debug-WK2-Tests-EWS "Waiting to run tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/3549 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/3083 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/3096 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/896 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/844 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/3106 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/3641 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/3361 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/1007 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

ddkilzer commented on Source/bmalloc/libpas/src/libpas/pas_enumerator.c:
```diff
@@ -462,4 +464,92 @@ bool pas_enumerator_enumerate_all(pas_enumerator* enumerator)
+memory_reader_adapter(task_t task, vm_address_t address, vm_size_t size, void **local_memory)
+{
+    void *ptr = memory_reader(task, address, size);
+    *local_memory = ptr;
```
Can a `nullptr` be passed for `local_memory`?  If so, this needs a `nullptr` check:

```
    if (local_memory)
        *local_memory = ptr;
```

stwrt replied:
I'll add the check.

You could theoretically pass a nullptr.

ddkilzer commented on Source/bmalloc/libpas/src/libpas/pas_enumerator.c:
```diff
@@ -462,4 +464,92 @@ bool pas_enumerator_enumerate_all(pas_enumerator* enumerator)
+    return memory_reader_adapter;
+}
+
+kern_return_t pas_pgm_extract_crash_report(vm_address_t fault_address, mach_vm_address_t pas_dead_root, addr64_t pgm_zone, task_t task, pas_pgm_report *report, crash_reporter_memory_reader_t crm_reader)
```
This is might not compile on older OSes since this signature changed, so you'll have to provide a stub for older OSes with a different signature.

We actually could support PGM on older OSes since new WebKit ships to older macOS releases regularly—though we would also need to back-port other internal changes to support it.  So you could also just make this new code conditional on future OS releases.

stwrt replied:
Which signature changed?

This is a new function and shouldn't have anything relying on it from the analytics side yet.

stwrt replied:
Talked with David. This should not cause any signature changes 😀.

ddkilzer commented on Source/bmalloc/libpas/src/libpas/pas_enumerator.h:
```diff
@@ -28,6 +28,8 @@
 #include "pas_allocation_config.h"
 #include "pas_utils.h"
+#include <mach/vm_types.h>
+#include <mach/mach_types.h>
```
Nit: Fix order of includes to be alphabetical per WebKit style.

ddkilzer replied:
This didn't get fixed in the second patch?  Maybe you for got to `git add` the change locally before amending?

ddkilzer reviewed (changes requested):
I'm not qualified to review `libpas` code, but there are a couple suggestions for improvements, and builds need to be fixed before this can land, so I'm marking this with "Request changes".

ddkilzer commented on Source/bmalloc/libpas/src/libpas/pas_enumerator.c:
```diff
@@ -37,6 +37,8 @@
 #include "pas_enumerator_region.h"
 #include "pas_ptr_hash_set.h"
+#include "pas_ptr_hash_map.h"
+#include "pas_probabilistic_guard_malloc_allocator.h"
```
Sorry I didn't catch this before—these should also be alphabetized.

ddkilzer reviewed (commented):
Just a couple nits for header include alphabetizing.

As I mentioned before, I don't think I'm qualified to review `libpas` code, but otherwise this looks fine.

ddkilzer commented on Source/bmalloc/libpas/src/libpas/pas_report_crash.h:
```diff
@@ -0,0 +1,59 @@
+    size_t allocation_size;
+} pas_report_crash_pgm_report;
+
+kern_return_t pas_report_crash_extract_pgm_failure(vm_address_t fault_address, mach_vm_address_t pas_dead_root, task_t task, pas_report_crash_pgm_report *report, crash_reporter_memory_reader_t crm_reader);
```
I think we need to mark this function as exported (via `BEXPORT` macro) if we want the symbol to be available in `JavaScriptCore.framework` (and not stripped as dead code).

stwrt replied:
I am not sure we want to start taking bmalloc code and requiring for libpas at the moment.

Elliott and I were discussing adding code to JavaScriptCore that would call into libpas.

We hide libpas symbols now if you try to go through linking JSC.

emw-apple replied:
Right – as of c885def, JavaScriptCore hides all symbols from libpas (making libpas an effective implementation detail). This was done to better align the symbols exported from JavaScriptCore with the declarations visible in JavaScriptCore/WTF/bmalloc headers. (And _that_ work is in support of InstallAPI, see https://bugs.webkit.org/show_bug.cgi?id=254784.)

emw-apple replied:
I suggested that Brandon add SPI to JavaScriptCore that proxies through to this pas call, since that respects the "libpas is implementation-only" principle we are now enforcing via the linker.

emw-apple replied:
> I suggested that Brandon add SPI to JavaScriptCore that proxies through to this pas call,

This could also be implemented in bmalloc, since it's _only_ the libpas symbols that are hidden. (bmalloc already has clients outside of JavaScriptCore, such as WebCore and WebKit.)

ddkilzer reviewed (approved):
r=me, but probably need to add the `BEXPORT` macro as noted.

emw-apple commented on Source/JavaScriptCore/API/PASReportCrashPrivate.h:
```diff
@@ -0,0 +1,44 @@
+#define PASReportCrashPrivate_hpp
+
+#include <stdio.h>
+#import <JavaScriptCore/JavaScriptCore.h>
```
Probably better to import just the one header you need here (looks like it may be `<JavaScriptCore/JSExportMacros.h>`). Also, use `#include` since this isn't objective-c.

emw-apple commented on Source/JavaScriptCore/API/PASReportCrashPrivate.h:
```diff
@@ -0,0 +1,44 @@
+ */
+
+#ifndef PASReportCrashPrivate_hpp
+#define PASReportCrashPrivate_hpp
```
I believe WebKit style is to use `#pragma once`.

emw-apple commented on Source/JavaScriptCore/API/PASReportCrashPrivate.cpp:
naming nit: I typically only see "Private" in the name when it's a private counterpart to public API. So I think `PASReportCrash.cpp` would be a fine rename.

stwrt replied:
Updated

emw-apple commented on Source/JavaScriptCore/API/PASReportCrashPrivate.cpp:
```diff
@@ -0,0 +1,34 @@
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <config.h>
```
I'm kind of surprised this works. config.h is usually a quote-style include.
```suggestion
#include "config.h"
```

emw-apple commented on Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
```diff
@@ -3,7 +3,7 @@
 	classes = {
 	};
-	objectVersion = 52;
+	objectVersion = 54;
```
I don't think we can bump this yet.
```suggestion
	objectVersion = 52;
```

emw-apple commented on Source/bmalloc/bmalloc.xcodeproj/project.pbxproj:
```diff
@@ -1290,6 +1292,9 @@
 		14F271BE18EA3963008C152F /* libbmalloc.a */ = {isa = PBXFileReference; explicitFileType = archive.ar; includeInIndex = 0; path = libbmalloc.a; sourceTree = BUILT_PRODUCTS_DIR; };
+		2BDF4F4A29E8B8BA0056BF50 /* pas_report_crash.c */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.c.c; name = pas_report_crash.c; path = libpas/src/libpas/pas_report_crash.c; sourceTree = "<group>"; };
+		2BDF4F4B29E8B8BA0056BF50 /* pas_report_crash.h */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.c.h; name = pas_report_crash.h; path = libpas/src/libpas/pas_report_crash.h; sourceTree = "<group>"; };
+		2BDF4F5A29E9A4A90056BF50 /* pas_report_crash.h */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.c.h; name = pas_report_crash.h; path = libpas/src/libpas/pas_report_crash.h; sourceTree = "<group>"; };
```
These look like duplicate references to the same file.

emw-apple commented on Source/bmalloc/libpas/libpas.xcodeproj/project.pbxproj:
```diff
@@ -552,6 +552,9 @@
 		2B6055E42805368B00C8BDAC /* BmallocTests.cpp in Sources */ = {isa = PBXBuildFile; fileRef = 2B6055E32805368B00C8BDAC /* BmallocTests.cpp */; };
+		2BDF4F4529E8B36F0056BF50 /* pas_report_crash.h in Headers */ = {isa = PBXBuildFile; fileRef = 2BDF4F4429E8B36F0056BF50 /* pas_report_crash.h */; };
+		2BDF4F4729E8B3AD0056BF50 /* pas_report_crash.c in Sources */ = {isa = PBXBuildFile; fileRef = 2BDF4F4629E8B3AD0056BF50 /* pas_report_crash.c */; };
+		2BDF4F4929E8B5510056BF50 /* pas_report_crash.c in Sources */ = {isa = PBXBuildFile; fileRef = 2BDF4F4829E8B5510056BF50 /* pas_report_crash.c */; };
```
Ditto duplicates.

emw-apple commented on Source/JavaScriptCore/API/PASReportCrashPrivate.h:
```diff
@@ -0,0 +1,47 @@
+extern "C" {
+#endif
+
+#if !PLATFORM(WATCHOS)
```
This is a different guard than what libpas uses:
```
#if PAS_OS(DARWIN)
```

Is that ok? It seems like non-Darwin ports would compile this function when `pas_report_crash_extract_pgm_failure` is not available.

stwrt replied:
I think we need both.

WATCHOS uses bmalloc

emw-apple commented on Source/JavaScriptCore/API/PASReportCrash.h:
```diff
@@ -0,0 +1,43 @@
+extern "C" {
+#endif
+
+#if !PLATFORM(WATCHOS) && PAS_OS(DARWIN)
```
PAS_OS probably shouldn't be used outside of libpas. `OS(DARWIN)` from WTF would be better here.

emw-apple commented on Source/JavaScriptCore/API/PASReportCrash.h:
More thoughts on the ifdefs: I think you mean to compile this on Darwin-based platforms that use libpas (i.e. Darwin platforms minus watchOS). We can be more precise than conditionalizing on platform. How about:
```
#if !USE(SYSTEM_MALLOC)
#include <bmalloc/BPlatform.h>
#if BENABLE(LIBPAS) && OS(DARWIN)

// code

#endif
#endif
```

This is very simlar to what we have in: https://github.com/WebKit/WebKit/blob/a132303fd6a0c4b8ee96b6cc986723477ffefb98/Source/JavaScriptCore/jit/ExecutableMemoryHandle.h#L37-L42

The USE(SYSTEM_MALLOC) check means that ports which don't use bmalloc at all will skip over this file. Every other include of <bmalloc/…> in JSC is ifdef'd this way.

stwrt replied:
Sure will change to `BENABLE(LIBPAS) && OS(DARWIN)`

stwrt replied:
Should be fixed now

emw-apple reviewed (approved):
r=me with feedback. I recommend getting a JSC reviewer to rubber-stamp the SPI you're adding and confirm that it's in the right subdirectory.

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_report_crash.h:
```diff
@@ -0,0 +1,59 @@
+    const char *confidence;
+    vm_address_t fault_address;
+    size_t allocation_size;
+} pas_report_crash_pgm_report;
```
Since this struct is being exposed as part of the part of the JSC SPI, I would have preferred that it be define in the JSC SPI header file.  However, that would introduce a layer violation.  I suppose we can live with this.

More importantly, we should define a version field at the top of the struct.  `pas_report_crash_extract_pgm_failure` should fill in this `version` so that the client can check if it knows how to parse the contents of `pas_report_crash_pgm_report`.

stwrt replied:
Added version as part of the function

MenloDorian commented on Source/JavaScriptCore/API/PASReportCrash.h:
```diff
@@ -0,0 +1,43 @@
+#endif
+
+#if BENABLE(LIBPAS) && OS(DARWIN)
+JS_EXPORT_PRIVATE kern_return_t PASReportCrashExtractResults(vm_address_t fault_address, mach_vm_address_t pas_dead_root, task_t task, pas_report_crash_pgm_report *report, crash_reporter_memory_reader_t crm_reader);
```
You should use `JS_EXPORT` (like other API and SPI functions) instead of `JS_EXPORT_PRIVATE`.

The declaration need to have `JSC_API_AVAILABLE(macos(JSC_MAC_TBA), ios(JSC_IOS_TBA)) at the end of it.

There's tension here with this being libpas SPI and not JSC SPI.  The only reason this file is in the JSC API directory is because it is packaged as part of the JSC framework.  With that ...

It seems odd that we're working with a `pas_report_crash_pgm_report` struct follows libpas style while the function name follows JSC style.  How about we just call this function `pas_report_crash_extract_results` instead and go all in on being consistent with libpas style rather than being halfway there.

Also let's call this file `PasReportCrashPrivate.h` to indicate that it is SPI.  It should not be included in the public SDK.

MenloDorian replied:
Also, this SPI should take a `size_t reportSize` argument after the `report` argument.  See `pas_report_crash_extract_pgm_failure` for more details.

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_report_crash.h:
```diff
@@ -0,0 +1,59 @@
+PAS_BEGIN_EXTERN_C;
+
+#if PAS_OS(DARWIN)
+// Read memory from crashed process.
```
Please add a comment somewhere at the top if this fils to indicate that this file defines SPI for ReportCrash, and that we should not be adding non-API/SPI content in this file.

stwrt replied:
Added a comment

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_report_crash.h:
```diff
@@ -0,0 +1,59 @@
+    size_t allocation_size;
+} pas_report_crash_pgm_report;
+
+PAS_API kern_return_t pas_report_crash_extract_pgm_failure(vm_address_t fault_address, mach_vm_address_t pas_dead_root, task_t task, pas_report_crash_pgm_report *report, crash_reporter_memory_reader_t crm_reader);
```
This should take a `size_t reportSize` argument after the `report` argument.  `pas_report_crash_extract_pgm_failure` should check that `reportSize` is >= sizeof(pas_report_crash_pgm_report)`.  Otherwise, fail.

MenloDorian replied:
Does this still need to be PAS_API?  `pas_report_crash_extract_results` is now the SPI.  This is just an internal implementation function.

stwrt replied:
I don't agree including a report size.

If the versions number mismatch we should abandon as we can't guarantee how to parse it anymore. Version numbers may change the content without changing the size.

MenloDorian reviewed (changes requested):
I think this patch is not ready to land as is.

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_report_crash.h:
```diff
@@ -0,0 +1,59 @@
+
+#include "pas_ptr_hash_set.h"
+#include "pas_ptr_hash_map.h"
+#include "pas_root.h"
```
Looks like these aren't needed in the header file.  Just include them in the ,c file.

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_report_crash.h:
```diff
@@ -0,0 +1,59 @@
+    const char *confidence;
+    vm_address_t fault_address;
+    size_t allocation_size;
+} pas_report_crash_pgm_report;
```
Let's put `crash_reporter_memory_reader_t` and `pas_report_crash_pgm_report` in a `pas_report_crash_pgm_report.h` file (which doesn't #include lots of implementation details and is safe to include from a SPI header).

In that header file, don't including pas_util.h and all the Darwin headers.  Don't use all the PAS macros.  Instead, do:
```
#if defined(__has_include) && __has_include(<mach/vm_types.h>)
#include <mach/vm_types.h>
#else
#define vm_address_t void*
#endif
```
 
Instead of using PAS_BEGIN_EXTERN_C, just use `extern "C".  We don't want an external project dependent on libpas or WebKit's build configurations.

MenloDorian commented on Source/JavaScriptCore/API/PASReportCrash.h:
```diff
@@ -0,0 +1,43 @@
+extern "C" {
+#endif
+
+#if BENABLE(LIBPAS) && OS(DARWIN)
```
Remove this `#if`.  Make it unconditional.  We don't want an external project dependent on WebKit's build configuration settings.  The function should always exist and just return a failure status if libpas is not enabled, or if this is linked on a non-Darwin platform.  Let's reduce API/SPI dependencies on all these WebKit internal macros.

MenloDorian commented on Source/JavaScriptCore/API/PASReportCrash.h:
```diff
@@ -0,0 +1,43 @@
+
+#include <stdio.h>
+#include <JavaScriptCore/JSExportMacros.h>
+#include <bmalloc/pas_report_crash.h>
```
Just include `pas_report_crash_pgm_report.h` instead.

MenloDorian replied:
In this file, also add:
```
#if defined(__has_include) && __has_include(<mach/vm_types.h>)
#include <mach/mach_types.h>
#include <mach/vm_types.h>
#else
#define kern_return_t unsigned
#define mach_vm_address_t void*
#define vm_address_t void*
#define crash_reporter_memory_reader_t void*
#endif
```

MenloDorian commented on Source/JavaScriptCore/API/PASReportCrash.cpp:
```diff
@@ -0,0 +1,36 @@
+
+using namespace JSC;
+
+#if BENABLE(LIBPAS) && OS(DARWIN)
```
Move this `#if` condition inside the function.  If the condition is not true, make the function return a failure status.

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_report_crash.h:
```diff
@@ -0,0 +1,61 @@
+/* This file exposes an SPI between OSAnalytics and libpas ultimately called through
+ * JavaScriptCore. Upon crashing of a process on Apple platforms ReportCrash will call
+ * into this function to inspect whether it was a PGM crash in libpas or not. We will report
+ * back results to libpas with any information about the PGM crash. This will be logged in
```
typo: /to libpas/from libpas/.

stwrt replied:
Fixed

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_report_crash.h:
```diff
@@ -0,0 +1,61 @@
+#ifdef __APPLE__
+#include <mach/mach_types.h>
+#include <mach/vm_types.h>
+
```
nit: Can you move the above comment (line 26-30) here?

stwrt replied:
Fixed

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_report_crash.h:
```diff
@@ -0,0 +1,61 @@
+// Read memory from crashed process.
+typedef void *(*crash_reporter_memory_reader_t)(task_t task, vm_address_t address, size_t size);
+
+const size_t pas_crash_report_version = 1;
```
nit: it feels strange that the type for a "version" is `size_t` which implies it's some sort of size.  Why not use `unsigned`?

stwrt replied:
Sure

ddkilzer commented on Source/JavaScriptCore/API/PASReportCrashPrivate.h:
```diff
@@ -0,0 +1,44 @@
+#include <JavaScriptCore/JSBase.h>
+
+#ifdef __APPLE__
+#include <TargetConditionals.h>
```
The `<TargetConditions.h>` header can likely be moved to `PASReportCrashPrivate.cpp`.

stwrt replied:
Done

ddkilzer reviewed (approved):
r=me with a minor suggestion to move a `#include` statement.

MenloDorian commented on Source/bmalloc/libpas/src/pas_report_crash_pgm_report.h:
```diff
@@ -0,0 +1,59 @@
+
+const unsigned pas_crash_report_version = 1;
+
+// Report sent back to the ReportCrash process.
```
Let's make this a C-style comment since this is meant to be a C header file.

stwrt replied:
Sure

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_report_crash.h:
```diff
@@ -0,0 +1,50 @@
+#endif
+
+// Read memory from crashed process.
+typedef void *(*crash_reporter_memory_reader_t)(task_t task, vm_address_t address, size_t size);
```
This belongs in `pas_report_crash_pgm_report.h` because it is part of the SPI that external clients call.  `PASReportCrashExtractResults` takes it as an argument, right?

MenloDorian commented on Source/JavaScriptCore/API/PASReportCrashPrivate.h:
```diff
@@ -0,0 +1,43 @@
+#include <JavaScriptCore/JSBase.h>
+
+#ifdef __APPLE__
+#include <bmalloc/pas_report_crash.h>
```
This should include `pas_report_crash_pgm_report.h` instead of `pas_report_crash.h`.

MenloDorian reviewed (approved):
r=me with the 3 remaining issues fixed.

stwrt committed:
PGM Crash Analysis
https://bugs.webkit.org/show_bug.cgi?id=255345

Reviewed by Mark Lam, Elliott Williams and David Kilzer.

When a process crashes we will now try to determine whether this crash was caused by
a PGM allocation in WebKit. ReportCrash will call into JSC, which would forward the information
onto libpas. libpas will respond with a report generated from the memory of the now dead process.

libpas will determine whether this was a PGM crash and if so what kind of crash (UAF or OOB).
This information will be added to the local crash log generated.

* Source/JavaScriptCore/API/PASReportCrashPrivate.cpp: Added.
(PASReportCrashExtractResults):
* Source/JavaScriptCore/API/PASReportCrashPrivate.h: Added.
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/SourcesCocoa.txt:
* Source/bmalloc/CMakeLists.txt:
* Source/bmalloc/bmalloc.xcodeproj/project.pbxproj:
* Source/bmalloc/libpas/libpas.xcodeproj/project.pbxproj:
* Source/bmalloc/libpas/src/libpas/pas_enumerate_large_heaps.c:
(pas_hash_map_entry_callback):
* Source/bmalloc/libpas/src/libpas/pas_report_crash.c: Added.
(memory_reader_adapter):
(setup_memory_reader):
(pas_report_crash_extract_pgm_failure):
* Source/bmalloc/libpas/src/libpas/pas_report_crash.h: Added.
* Source/bmalloc/libpas/src/libpas/pas_report_crash_pgm_report.h: Added.

Canonical link: https://commits.webkit.org/263055@main
